src/MobileDevice/widget/MobileDevice.ts

```diff
@@ -0,0 +1,151 @@
+import * as dojoDeclare from "dojo/_base/declare";
+import * as domConstruct from "dojo/dom-construct";
+import * as WidgetBase from "mxui/widget/_WidgetBase";
+import * as dojoLang from "dojo/_base/lang";
+import * as dojoOn from "dojo/on";
+
+type onDeviceReadyActions = "doNothing" | "showPage" | "callMicroflow";
+
+class MobileDevice extends WidgetBase {
+    private deviceIdAttribute: string;
+    private deviceTypeAttribute: string;
+    private appVersionVersionAttribute: string;
+    private appVersionBuildAttribute: string;
+    private microflow: string;
+    private onDeviceReadyAction: onDeviceReadyActions;
+    private page: string;
+    private mxObject: mendix.lib.MxObject;
+    private onNavigateBack: boolean;
+    // internal variables
+    private deviceReadyEvent: dojoEvent | null;
+
+    postCreate() {
+        this.showError(this.validateProps());
+        this.setUpWidgetDom();
+    }
+
+    update(mxObject: mendix.lib.MxObject, callback: () => void) {
+        this.mxObject = mxObject;
+        if (!window.device) {
+            mxObject.set(this.deviceIdAttribute, "");
+            mxObject.set(this.deviceTypeAttribute, "Web");
+            this.commit(mxObject);
+        }
+        this.setUpEvents();
+
+        callback();
+    }
+
+    uninitialize(): boolean {
+        this.removeEvents();
+
+        return true;
+    }
+
+    private showError(message: string) {
+        if (message) {
+            domConstruct.create("div", {
+                class: "alert alert-danger",
+                innerHTML: message
+            }, this.domNode);
+        }
+    }
+
+    private setUpEvents() {
+        if (window.device) {
+            this.deviceReadyEvent = dojoOn(document, "deviceReady", dojoLang.hitch(this, this.onDeviceReady));
+        }
+    }
+
+    private removeEvents() {
+        if (window.device) {
+            if (this.deviceReadyEvent) {
+                this.deviceReadyEvent.remove();
+                this.deviceReadyEvent = null;
+            }
+        }
+    }
+
+    private setUpWidgetDom() {
+        domConstruct.create("div", {
+            class: "mobile-device-widget",
+            id: "mobile-device"
+        }, document.body, "first");
+    }
+
+    private onDeviceReady() {
+        if (cordova.hasOwnProperty("getAppVersion")) {
+            cordova.getAppVersion.getVersionNumber((versionNumber: string) => {
+                if (versionNumber) {
+                    cordova.getAppVersion.getVersionCode((versionCode: string) => {
+                        if (versionCode) {
```

Comment:
In the implementation of the plugin there response of the values are in callback. my thinking is that on each request the device is queried. so the functions will only return the values once the device has responded asynchronously. so that is why i nest them.

Ideally all this information is supposed to exists in ` AppVersion.m` eg  AppVersion.name or  AppVersion.version etc.  see issue whiteoctober/cordova-plugin-app-version#14 this access is not stable. I also tried it my self and i was having issues so i went with the above implementation.